File: ui_dsl/stream_widgets.js
/* eslint-disable */
(function(){
  const sse = (topic, onMsg) => {
    const url = `/events?topic=${encodeURIComponent(topic)}`;
    const es = new EventSource(url);
    es.addEventListener('msg', ev => {
      try {
        const data = JSON.parse(ev.data);
        onMsg && onMsg(data);
      } catch(e){}
    });
    es.onerror = () => {}; // keep-alive by server
    return es;
  };

  class StreamProgress extends HTMLElement {
    connectedCallback(){
      this.attachShadow({mode:'open'});
      this.shadowRoot.innerHTML = `
        <style>
          .bar { height: 8px; background:#eee; border-radius:4px; }
          .fill { height: 8px; background:#4b7bec; width:0%; transition: width .2s; border-radius:4px; }
          .txt { font:12px system-ui, sans-serif; color:#333; margin-top:4px; }
        </style>
        <div class="bar"><div class="fill"></div></div>
        <div class="txt">waiting…</div>
      `;
      const fill = this.shadowRoot.querySelector('.fill');
      const txt  = this.shadowRoot.querySelector('.txt');
      const topic = this.getAttribute('topic') || 'progress';
      let cur=0, total=1;
      this._es = sse(topic, (m)=>{
        if(m.stage==='adapter_start'){ cur = (m.i||1)-1; total = m.n||1; }
        if(m.stage==='adapter_done'){ cur = m.i||1; total = m.n||1; }
        if(m.stage==='complete'){ cur = total; }
        const pct = Math.max(0, Math.min(100, Math.floor(100*cur/Math.max(1,total))));
        fill.style.width = pct + '%';
        txt.textContent = `${pct}% – ${m.stage||''} ${(m.kind||'')}`;
      });
    }
    disconnectedCallback(){
      if(this._es) this._es.close();
    }
  }
  customElements.define('stream-progress', StreamProgress);

  class EventTimeline extends HTMLElement {
    connectedCallback(){
      this.attachShadow({mode:'open'});
      this.shadowRoot.innerHTML = `
        <style>
          ul { list-style:none; padding:0; margin:0; font:12px system-ui, sans-serif; }
          li { padding:6px 8px; border-bottom:1px solid #eee; }
          .kind { color:#999; margin-left:6px; }
        </style>
        <ul></ul>
      `;
      const ul = this.shadowRoot.querySelector('ul');
      const topic = this.getAttribute('topic') || 'timeline';
      this._es = sse(topic, (m)=>{
        const li = document.createElement('li');
        const t = m.t || m.stage || 'evt';
        let line = `[${new Date().toLocaleTimeString()}] ${t}`;
        if(m.kind) line += ` · ${m.kind}`;
        if(m.err)  line += ` · ❌ ${m.err}`;
        li.textContent = line;
        ul.prepend(li);
        while(ul.children.length > 200) ul.removeChild(ul.lastChild);
      });
    }
    disconnectedCallback(){ if(this._es) this._es.close(); }
  }
  customElements.define('event-timeline', EventTimeline);
})();

// Lightweight client runtime (vanilla JS)
export function mountProgressBar(el, topic, broker) {
  let val = 0;
  const bar = document.createElement('div');
  bar.style.height = '8px';
  bar.style.background = '#eee';
  const fill = document.createElement('div');
  fill.style.height = '8px';
  fill.style.width = '0%';
  fill.style.background = '#4caf50';
  bar.appendChild(fill);
  el.appendChild(bar);
  broker.subscribe(topic, (evt) => {
    if (typeof evt.progress === 'number') {
      val = Math.max(0, Math.min(100, evt.progress));
      fill.style.width = val + '%';
    }
  });
}

export function mountEventTimeline(el, topic, broker) {
  const list = document.createElement('ul');
  list.style.listStyle = 'none';
  list.style.padding = '0';
  el.appendChild(list);
  broker.subscribe(topic, (evt) => {
    const li = document.createElement('li');
    li.textContent = `[${new Date().toISOString()}] ${evt.message || JSON.stringify(evt)}`;
    list.appendChild(li);
    if (list.childNodes.length > 200) list.removeChild(list.firstChild);
  });
}

export function mountDataTable(el, topic, broker, { freezeLeft=1 } = {}) {
  const table = document.createElement('table');
  table.style.borderCollapse = 'collapse';
  table.style.width = '100%';
  const thead = document.createElement('thead');
  const tbody = document.createElement('tbody');
  table.appendChild(thead); table.appendChild(tbody);
  el.appendChild(table);
  let columns = null;
  broker.subscribe(topic, (evt) => {
    if (evt.columns && !columns) {
      columns = evt.columns;
      const tr = document.createElement('tr');
      columns.forEach((c, idx) => {
        const th = document.createElement('th');
        th.textContent = c;
        th.style.position = idx < freezeLeft ? 'sticky' : 'static';
        th.style.left = idx < freezeLeft ? (idx*120)+'px' : '0';
        th.style.background = '#fff';
        th.style.borderBottom = '1px solid #ddd';
        th.style.padding = '4px 8px';
        tr.appendChild(th);
      });
      thead.appendChild(tr);
    }
    if (evt.row) {
      const tr = document.createElement('tr');
      evt.row.forEach((v, idx) => {
        const td = document.createElement('td');
        td.textContent = String(v);
        td.style.position = idx < freezeLeft ? 'sticky' : 'static';
        td.style.left = idx < freezeLeft ? (idx*120)+'px' : '0';
        td.style.background = idx < freezeLeft ? '#fafafa':'#fff';
        td.style.borderBottom = '1px solid #f0f0f0';
        td.style.padding = '4px 8px';
        tr.appendChild(td);
      });
      tbody.appendChild(tr);
      if (tbody.childNodes.length > 1000) tbody.removeChild(tbody.firstChild);
    }
  });
}

File: ui_dsl/stream_widgets.py
# ui_dsl/stream_widgets.py
# -*- coding: utf-8 -*-
import json, html
from typing import List, Dict, Any

# unique IDs for multiple widgets on the same page
_TL_COUNTER = 0

def _next_id(prefix: str) -> str:
    global _TL_COUNTER
    _TL_COUNTER += 1
    return f"{prefix}{_TL_COUNTER}"



def render_progress_bars(bars: List[Dict[str, Any]]) -> str:
    """
    Stream-friendly progress bars widget.
    bars: [{"id":"p1","label":"Encode","value":35}, ...]
    - Live updates on WS topic /^progress\// with {topic:"progress/<id>", value:Number}.
    - Accessible (ARIA) and self-contained CSS+JS.
    """
    s = ['<div class="imu-progress-area" role="group" aria-label="progress group">']
    for b in bars:
        _id = html.escape(b["id"])  # required
        _label = html.escape(b.get("label", ""))
        _val = int(b.get("value", 0))
        s.append(f'''
<div class="imu-progress" id="{_id}" role="progressbar" aria-valuemin="0" aria-valuemax="100" aria-valuenow="{_val}">
  <div class="imu-progress__label">{_label}</div>
  <div class="imu-progress__outer"><div class="imu-progress__inner" style="width:{_val}%"></div></div>
</div>
''')
    s.append("""
<style>
.imu-progress-area{display:grid;gap:8px}
.imu-progress__outer{background:#eee;height:10px;border-radius:6px;overflow:hidden}
.imu-progress__inner{height:10px;transition:width .15s}
.imu-progress__label{font:12px sans-serif;margin-bottom:2px}
</style>
<script>
(function(){
  'use strict';
  function setVal(id, v){
    const el = document.getElementById(id);
    if(!el) return;
    const val = (v|0);
    const inner = el.querySelector('.imu-progress__inner');
    if(inner) inner.style.width = val + '%';
    el.setAttribute('aria-valuenow', String(val));
  }
  const ws = window.IMU_WS;
  if(ws && typeof ws.subscribe === 'function'){
    ws.subscribe(/^progress\//, function(msg){
      // msg: {topic:"progress/<id>", value:Number}
      const t = (msg && msg.topic) ? String(msg.topic) : '';
      const id = t.split('/')[1];
      if(!id) return;
      setVal(id, msg.value);
    });
  }
})();
</script>
""")
    s.append("</div>")
    return "".join(s)


def render_event_timeline(events: List[Dict[str, Any]]) -> str:
    """
    Stream-ready timeline widget with filters, client-side sort, and efficient incremental updates.
    - Initial events are provided by `events`.
    - Live updates arrive on WS topic /^timeline\// (expects {topic, event{ts,type,text}}) or EventSource fallback.
    - Virtualizes to keep last MAX_ITEMS to avoid DOM bloat.
    - Safe against HTML injection by using textContent only.
    """
    safe = json.dumps(events, ensure_ascii=False).replace("</", "<\/")

    tl_id = _next_id("imuTL_")
    type_id = _next_id("fltType_")
    text_id = _next_id("fltText_")
    sort_id = _next_id("fltSortTs_")

    html_doc = """
<div class="imu-timeline">
  <div class="imu-timeline__filters">
    <label>Type:
      <select id="__TYPE__">
        <option value="">Any</option><option>info</option><option>warn</option><option>error</option>
      </select>
    </label>
    <label>Search: <input id="__TEXT__" placeholder="contains..."/></label>
    <button id="__SORT__">Sort by time</button>
  </div>
  <ul id="__TL__" class="imu-timeline__list"></ul>
</div>
<style>
.imu-timeline__list{list-style:none;padding:0;margin:8px 0}
.imu-timeline__list li{padding:6px 8px;border-bottom:1px solid #eee;font:13px sans-serif}
.imu-tag-info{color:#0a0}
.imu-tag-warn{color:#c80}
.imu-tag-error{color:#c00}
</style>
<script>
(function(){
  const MAX_ITEMS = 500;             // keep up to N rendered items
  const TOPIC_RE = /^timeline\//;    // which topics to accept
  const data = __SAFE__;
  let cur = data.slice();
  let asc = true;

  const el = document.getElementById('__TL__');
  const fType = document.getElementById('__TYPE__');
  const fText = document.getElementById('__TEXT__');
  const bSort = document.getElementById('__SORT__');

  function makeItem(ev){
    const li = document.createElement('li');
    const tag = document.createElement('span');
    const strong = document.createElement('b');

    const cls = ev.type==='error'?'imu-tag-error':(ev.type==='warn'?'imu-tag-warn':'imu-tag-info');
    tag.className = cls;
    tag.textContent = ev.type || 'info';

    const dt = new Date(ev.ts*1000).toISOString();
    strong.textContent = dt;

    li.appendChild(tag);
    li.appendChild(document.createTextNode(' '));
    li.appendChild(strong);
    li.appendChild(document.createTextNode(' — '));
    li.appendChild(document.createTextNode(ev.text || ''));
    return li;
  }

  function render(list){
    const frag = document.createDocumentFragment();
    const limit = Math.min(list.length, MAX_ITEMS);
    const view = asc ? list.slice(-limit) : list.slice(0, limit);
    el.textContent = '';
    if (asc){
      for (let i=0;i<view.length;i++) frag.appendChild(makeItem(view[i]));
    } else {
      for (let i=view.length-1;i>=0;i--) frag.appendChild(makeItem(view[i]));
    }
    el.appendChild(frag);
  }

  function apply(){
    const t = fType ? fType.value : '';
    const s = (fText && fText.value ? fText.value : '').toLowerCase();
    cur = data.filter(ev => (!t || ev.type===t) && (!s || (ev.text||'').toLowerCase().includes(s)));
    cur.sort((a,b)=> asc ? (a.ts-b.ts) : (b.ts-a.ts));
    render(cur);
  }

  if (fType) fType.onchange = apply;
  if (fText) fText.oninput = apply;
  if (bSort) bSort.onclick = ()=>{ asc=!asc; apply(); };
  apply();

  // --- streaming & batching ---
  const ws = window.IMU_WS;
  let queue = [];
  let scheduled = false;
  function scheduleFlush(){
    if (scheduled) return;
    scheduled = true;
    requestAnimationFrame(()=>{
      scheduled = false;
      if (!queue.length) return;
      // append and cap underlying data for memory safety
      for (let i=0;i<queue.length;i++) data.push(queue[i]);
      queue.length = 0;
      if (data.length > MAX_ITEMS*4) data.splice(0, data.length - MAX_ITEMS*2);
      apply();
    });
  }

  function onMsg(msg){
    // msg: {topic:"timeline/main", event:{ts,type,text}}
    if (!msg || !msg.topic || !TOPIC_RE.test(msg.topic)) return;
    const ev = msg.event;
    if (ev && typeof ev.ts === 'number'){
      queue.push(ev);
      scheduleFlush();
    }
  }

  if (ws && typeof ws.subscribe === 'function'){
    ws.subscribe(/^timeline\//, onMsg);
  } else if (window.IMU_STREAM_URL && window.EventSource){
    try{
      const es = new EventSource(window.IMU_STREAM_URL);
      es.onmessage = (e)=>{ try { onMsg(JSON.parse(e.data)); } catch(_){} };
    }catch(_){/* ignore */}
  }
})();
</script>
"""

    return (html_doc
            .replace("__SAFE__", safe)
            .replace("__TL__", tl_id)
            .replace("__TYPE__", type_id)
            .replace("__TEXT__", text_id)
            .replace("__SORT__", sort_id))
